src/prefect/task_engine.py

```diff
@@ -252,298 +218,777 @@ def _wait_for_dependencies(self):
+                    )
+
+                    self._is_started = False
+                    self._client = None
+
+    async def wait_until_ready(self):
+        """Waits until the scheduled time (if its the future), then enters Running."""
+        if scheduled_time := self.state.state_details.scheduled_time:
+            sleep_time = (scheduled_time - pendulum.now("utc")).total_seconds()
+            await anyio.sleep(sleep_time if sleep_time > 0 else 0)
+            self.set_state(
+                Retrying() if self.state.name == "AwaitingRetry" else Running(),
+                force=True,
+            )
+
+    def handle_crash(self, exc: BaseException) -> None:
+        state = run_coro_as_sync(exception_to_crashed_state(exc))
+        self._log_task_run_crashed(state, exc)
+        self.set_state(state, force=True)
+        self._raised = exc
+
+    # --------------------------
+    #
+    # The following methods compose the main task run loop
+    #
+    # --------------------------
+
+    @contextmanager
+    def start(
+        self,
+        task_run_id: Optional[UUID] = None,
+        dependencies: Optional[Dict[str, Set[TaskRunInput]]] = None,
+    ) -> Generator[None, None, None]:
+        with self.initialize_run(task_run_id=task_run_id, dependencies=dependencies):
+            self.begin_run()
+            try:
+                yield
+            finally:
+                self.call_hooks()
+
+    @contextmanager
+    def run_context(self):
+        result_factory = run_coro_as_sync(ResultFactory.from_task(self.task))
+
+        self.task_run = self.client.read_task_run(self.task_run.id)
+        self.set_task_run_name()
+
+        with self.setup_run_context(
+            result_factory=result_factory,
+            client=self.client,
+        ):
+            try:
+                with timeout(
+                    seconds=self.task.timeout_seconds,
+                    timeout_exc_type=TaskRunTimeoutError,
+                ):
+                    self.logger.debug(
+                        f"Executing task {self.task.name!r} for task run {self.task_run.name!r}..."
+                    )
+                    if self.is_cancelled():
+                        raise CancelledError("Task run cancelled by the task runner")
+
+                    yield self
+            except TimeoutError as exc:
+                self.handle_timeout(exc)
+            except Exception as exc:
+                self.handle_exception(exc)
+
+    def call_task_fn(self, transaction: Transaction) -> R:
+        """
+        Convenience method to call the task function.
+        """
+        parameters = self.parameters or {}
+        if transaction.is_committed():
+            result = transaction.read()
+        else:
+            result = call_with_parameters(self.task.fn, parameters)
+        self.handle_success(result, transaction=transaction)
+        return result
+
+
+@dataclass
+class AsyncTaskRunEngine(BaseTaskRunEngine[P, R]):
```

Comment:
We got lucky with the diff in general, because the way it reads here on Github, long stretches of it look like you're converting the sync engine to async and it's really helpful to reviewing it :D